src/validation/property-validator-schema.js

```diff
@@ -0,0 +1,260 @@
+  string: joi.string(),
+  integer: integerSchema,
+  float: joi.number(),
+  boolean: joi.boolean(),
+  datetime: datetimeStringSchema,
+  date: dateOnlyStringSchema,
+  enum: joi.alternatives().try([ joi.string(), integerSchema ]),
+  uuid: uuidSchema,
+  attachmentReference: joi.string(),
+  array: joi.array(),
+  object: joi.object().unknown(),
+  hashtable: joi.object().unknown()
+};
+
+var validPropertyTypes = Object.keys(typeEqualitySchemas).map(function(key) { return key; });
+
+var schema = joi.object().keys({
+  type: dynamicConstraintSchema(joi.string().only(validPropertyTypes)).required()
+})
+  .when(
+    joi.object().unknown().keys({ type: 'string' }),
+    { then: makeTypeConstraintsSchema('string') })
+  .when(
+    joi.object().unknown().keys({ type: 'integer' }),
+    { then: makeTypeConstraintsSchema('integer') })
+  .when(
+    joi.object().unknown().keys({ type: 'float' }),
+    { then: makeTypeConstraintsSchema('float') })
+  .when(
+    joi.object().unknown().keys({ type: 'boolean' }),
+    { then: makeTypeConstraintsSchema('boolean') })
+  .when(
+    joi.object().unknown().keys({ type: 'datetime' }),
+    { then: makeTypeConstraintsSchema('datetime') })
+  .when(
+    joi.object().unknown().keys({ type: 'date' }),
+    { then: makeTypeConstraintsSchema('date') })
+  .when(
+    joi.object().unknown().keys({ type: 'enum' }),
+    { then: makeTypeConstraintsSchema('enum') })
+  .when(
+    joi.object().unknown().keys({ type: 'uuid' }),
+    { then: makeTypeConstraintsSchema('uuid') })
+  .when(
+    joi.object().unknown().keys({ type: 'attachmentReference' }),
+    { then: makeTypeConstraintsSchema('attachmentReference') })
+  .when(
+    joi.object().unknown().keys({ type: 'array' }),
+    { then: makeTypeConstraintsSchema('array') })
+  .when(
+    joi.object().unknown().keys({ type: 'object' }),
+    { then: makeTypeConstraintsSchema('object') })
+  .when(
+    joi.object().unknown().keys({ type: 'hashtable' }),
+    { then: makeTypeConstraintsSchema('hashtable') })
+  .when(
+    joi.object().unknown().keys({ type: joi.func() }),
+    { then: joi.object().unknown() });
+
+/**
+ * A partial schema for a single entry in a "propertyValidators" object at either the top level of a document definition
+ * or nested within an "object" validator.
+ */
+module.exports = exports = schema;
+
+// Defined as a function rather than a plain object because it contains lazy references that result in recursive
+// references between the complex types (e.g. "array", "object", "hashtable") and the main "propertyValidators" schema
+function typeSpecificConstraintSchemas() {
+  return {
+    string: {
+      mustNotBeEmpty: dynamicConstraintSchema(joi.boolean()),
+      regexPattern: dynamicConstraintSchema(regexSchema),
+      minimumLength: dynamicConstraintSchema(integerSchema.min(0)),
+      maximumLength: maximumSizeConstraintSchema('minimumLength'),
+      minimumValue: dynamicConstraintSchema(joi.string()),
```

Comment:
Technically the constraints are allowed on strings (and any other data type). I figured it doesn't hurt to allow them in order to define a range of valid string values. For example `minimumValue: 'A', maximumValueExclusive: 'G' would allow only strings that start with the letters "A" through "F". I'll document them in the README.

Reply:
👍